src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java

```diff
@@ -39,4 +63,94 @@ public WarehouseDto save(WarehouseDto dto) {
     public void delete(Long id) {
 
     }
+
+    @Override
+    @Transactional
+    public void acceptItems(final ApplicationDto appDto, final Long warehouseId) {
+        lock.lock();
+        try {
+            applicationService.changeStatus(appDto.getId(), ApplicationStatus.STARTED_PROCESSING);
+            final ApplicationDto appFromDb = applicationService.findById(appDto.getId());
+            final Long destinationLocationId = appFromDb.getDestinationLocationDto().getId();
+            if (!destinationLocationId.equals(warehouseId)) {
+                throw new AccessDeniedException("Application doesn't belong to the requested warehouse");
+            }
+            final Long appId = appFromDb.getId();
+            final Set<ItemsInApplicationDto> acceptedToWarehouse = getItemsToAccept(appDto.getItems(), appId);
+
+            checkWarehouseCapacity(acceptedToWarehouse, destinationLocationId);
+            acceptToWarehouse(acceptedToWarehouse, destinationLocationId);
+            if (applicationService.isApplicationFullySatisfied(appId)) {
+                applicationService.changeStatus(appId, ApplicationStatus.FINISHED_PROCESSING);
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public Double getAvailableCapacity(final Long warehouseId) {
+        final Double availableCapacity = itemInWarehouseRepository.getAvailableCapacity(warehouseId);
+        return Objects.isNull(availableCapacity) ?
+                warehouseRepository.getTotalCapacity(warehouseId)
+                : availableCapacity;
+    }
+
+    private Set<ItemsInApplicationDto> getItemsToAccept(final Set<ItemsInApplicationDto> itemToAccept,
+                                                        final Long appId) {
+        final List<Long> acceptedItemInAppId = itemToAccept
+                .stream()
+                .map(ItemsInApplicationDto::getId)
+                .collect(Collectors.toList());
+        final Set<ItemsInApplicationDto> acceptedItemFromDb = applicationService
+                .getItemsById(acceptedItemInAppId, appId);
+        if (acceptedItemFromDb.size() != itemToAccept.size()) {
+            throw new ConflictWithTheCurrentStateException("Attempt to accept doesn't exist or already accepted item");
+        }
+        return acceptedItemFromDb;
+    }
+
+    private void acceptToWarehouse(final Set<ItemsInApplicationDto> acceptedToWarehouse,
+                                   final Long warehouseId) {
+        final List<Long> acceptedIds = new ArrayList<>();
+        acceptedToWarehouse.forEach(itemToAccept -> {
```

Comment:
use for each construction to avoid side effects in functional programming approach